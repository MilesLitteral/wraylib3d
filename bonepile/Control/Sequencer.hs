module Manifest.Control.Sequencer where

import Data.Aeson

data EventType 
    = SET_LOW -- UNTRIGGER
    |SET_HIGH -- TRIGGER
    deriving (Show, Eq, Enum)

data Sequence =
    Sequence {
        stepDuration :: Float,
        seqLen       :: Int --seqSlices    :: [SequenceSlice]
} deriving (Show, Eq)

data SequenceEvent = 
    SequenceEvent {
        groupIndex     :: Int,
        deviceIndex    :: Int,
        state          :: EventType
    } deriving (Show, Eq)

-- Represents a vertical slice of one step in a sequence
-- There is one BankState per DeviceBank
-- data SequenceSlice =
--     SequenceSlice {
--     bankStates :: [BankState]  
-- } deriving (Show, Eq)

-- class IsSequence a where 
--     executeSequence :: a -> Bool;

-- instance ToJSON SequenceEvent where
-- toJSON s = object
--     [  "sequence_index"  .=    sequenceIndex s,
--         "bank_index"     .=    bankIndex     s,
--         "device_index"   .=    deviceIndex   s,
--         "event"          .=    event s
--     ]

-- instance FromJSON SequenceEvent where
--     parseJSON (Object v) = SequenceEvent
--                             <$> v .: "sequence_index"
--                             <*> v .: "bank_index"
--                             <*> v .: "device_index"
--                             <*> v .: "event"


-- eventsToSlices :: [SequenceEvent] -> [DeviceBank] -> Int -> [SequenceSlice] 
-- eventsToSlices events banks numSteps = do
--     undefinied

-- executeSequence :: SRBoard -> Bool
-- executeSequence sr = undefined
