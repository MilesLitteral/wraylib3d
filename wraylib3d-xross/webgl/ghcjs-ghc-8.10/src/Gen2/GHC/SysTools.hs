-- partial copy

module Gen2.GHC.SysTools where

import Module
import Packages
import Config
import Panic
import Platform
import DynFlags

import Data.List (isPrefixOf, isInfixOf)

import System.FilePath
import System.Directory

import qualified SysTools as ST


linkDynLib :: DynFlags -> [String] -> [PackageId] -> IO ()
linkDynLib dflags0 o_files dep_packages
 = do
    let -- This is a rather ugly hack to fix dynamically linked
        -- GHC on Windows. If GHC is linked with -threaded, then
        -- it links against libHSrts_thr. But if base is linked
        -- against libHSrts, then both end up getting loaded,
        -- and things go wrong. We therefore link the libraries
        -- with the same RTS flags that we link GHC with.
        dflags1 = if cGhcThreaded then addWay' WayThreaded dflags0
                                  else                     dflags0
        dflags2 = if cGhcDebugged then addWay' WayDebug dflags1
                                  else                  dflags1
        dflags = updateWays dflags2

        verbFlags = getVerbFlags dflags
        o_file = outputFile dflags

    pkgs <- getPreloadPackagesAnd dflags dep_packages

    let pkg_lib_paths = collectLibraryPaths pkgs
    let pkg_lib_path_opts = concatMap get_pkg_lib_path_opts pkg_lib_paths
        get_pkg_lib_path_opts l
         | osElfTarget (platformOS (targetPlatform dflags)) &&
           dynLibLoader dflags == SystemDependent &&
           not (gopt Opt_Static dflags)
            = ["-L" ++ l, "-Wl,-rpath", "-Wl," ++ l]
         | otherwise = ["-L" ++ l]

    let lib_paths = libraryPaths dflags
    let lib_path_opts = map ("-L"++) lib_paths

    -- We don't want to link our dynamic libs against the RTS package,
    -- because the RTS lib comes in several flavours and we want to be
    -- able to pick the flavour when a binary is linked.
    -- On Windows we need to link the RTS import lib as Windows does
    -- not allow undefined symbols.
    -- The RTS library path is still added to the library search path
    -- above in case the RTS is being explicitly linked in (see #3807).
    let platform = targetPlatform dflags
        os = platformOS platform
        pkgs_no_rts = case os of
                      OSMinGW32 ->
                          pkgs
                      _ ->
                          filter ((/= rtsPackageId) . packageConfigId) pkgs
    let pkg_link_opts = let (package_hs_libs, extra_libs, other_flags) = ghcjsCollectLinkOpts dflags pkgs_no_rts
                        in  package_hs_libs ++ extra_libs ++ other_flags

        -- probably _stub.o files
    let extra_ld_inputs = ldInputs dflags
    case os of
        OSMinGW32 -> do
            -------------------------------------------------------------
            -- Making a DLL
            -------------------------------------------------------------
            let output_fn = case o_file of
                            Just s -> s
                            Nothing -> "HSdll.dll"

            ST.runLink dflags (
                    map Option verbFlags
                 ++ [ Option "-o"
                    , FileOption "" output_fn
                    , Option "-shared"
                    ] ++
                    [ FileOption "-Wl,--out-implib=" (output_fn ++ ".a")
                    | gopt Opt_SharedImplib dflags
                    ]
                 ++ map (FileOption "") o_files

                 -- Permit the linker to auto link _symbol to _imp_symbol
                 -- This lets us link against DLLs without needing an "import library"
                 ++ [Option "-Wl,--enable-auto-import"]

                 ++ extra_ld_inputs
                 ++ map Option (
                    lib_path_opts
                 ++ pkg_lib_path_opts
                 ++ pkg_link_opts
                ))
        OSDarwin -> do
            -------------------------------------------------------------------
            -- Making a darwin dylib
            -------------------------------------------------------------------
            -- About the options used for Darwin:
            -- -dynamiclib
            --   Apple's way of saying -shared
            -- -undefined dynamic_lookup:
            --   Without these options, we'd have to specify the correct
            --   dependencies for each of the dylibs. Note that we could
            --   (and should) do without this for all libraries except
            --   the RTS; all we need to do is to pass the correct
            --   HSfoo_dyn.dylib files to the link command.
            --   This feature requires Mac OS X 10.3 or later; there is
            --   a similar feature, -flat_namespace -undefined suppress,
            --   which works on earlier versions, but it has other
            --   disadvantages.
            -- -single_module
            --   Build the dynamic library as a single "module", i.e. no
            --   dynamic binding nonsense when referring to symbols from
            --   within the library. The NCG assumes that this option is
            --   specified (on i386, at least).
            -- -install_name
            --   Mac OS/X stores the path where a dynamic library is (to
            --   be) installed in the library itself.  It's called the
            --   "install name" of the library. Then any library or
            --   executable that links against it before it's installed
            --   will search for it in its ultimate install location.
            --   By default we set the install name to the absolute path
            --   at build time, but it can be overridden by the
            --   -dylib-install-name option passed to ghc. Cabal does
            --   this.
            -------------------------------------------------------------------

            let output_fn = case o_file of { Just s -> s; Nothing -> "a.out"; }

            instName <- case dylibInstallName dflags of
                Just n -> return n
                Nothing -> do
                    pwd <- getCurrentDirectory
                    return $ pwd `combine` output_fn
            ST.runLink dflags (
                    map Option verbFlags
                 ++ [ Option "-dynamiclib"
                    , Option "-o"
                    , FileOption "" output_fn
                    ]
                 ++ map Option o_files
                 ++ [ Option "-undefined",
                      Option "dynamic_lookup",
                      Option "-single_module" ]
                 ++ (if platformArch platform == ArchX86_64
                     then [ ]
                     else [ Option "-Wl,-read_only_relocs,suppress" ])
                 ++ [ Option "-install_name", Option instName ]
                 ++ map Option lib_path_opts
                 ++ extra_ld_inputs
                 ++ map Option pkg_lib_path_opts
                 ++ map Option pkg_link_opts
              )
        OSiOS -> throwGhcExceptionIO (ProgramError "dynamic libraries are not supported on iOS target")
        _ -> do
            -------------------------------------------------------------------
            -- Making a DSO
            -------------------------------------------------------------------

            let output_fn = case o_file of { Just s -> s; Nothing -> "a.out"; }
            let buildingRts = thisPackage dflags == rtsPackageId
            let bsymbolicFlag = if buildingRts
                                then -- -Bsymbolic breaks the way we implement
                                     -- hooks in the RTS
                                     []
                                else -- we need symbolic linking to resolve
                                     -- non-PIC intra-package-relocations
                                     ["-Wl,-Bsymbolic"]

            ST.runLink dflags (
                    map Option verbFlags
                 ++ [ Option "-o"
                    , FileOption "" output_fn
                    ]
                 ++ map Option o_files
                 ++ [ Option "-shared" ]
                 ++ map Option bsymbolicFlag
                    -- Set the library soname. We use -h rather than -soname as
                    -- Solaris 10 doesn't support the latter:
                 ++ [ Option ("-Wl,-h," ++ takeFileName output_fn) ]
                 ++ map Option lib_path_opts
                 ++ extra_ld_inputs
                 ++ map Option pkg_lib_path_opts
                 ++ map Option pkg_link_opts
              )

ghcjsCollectLinkOpts :: DynFlags -> [PackageConfig] -> ([String], [String], [String])
ghcjsCollectLinkOpts dflags ps =
    (
        concatMap (map ("-l" ++) . ghcjsPackageHsLibs dflags) ps,
        concatMap (map ("-l" ++) . extraLibraries) ps,
        concatMap ldOptions ps
    )

--------------------------------------------------
-- GHC API gets libs like libHSpackagename-ghcversion
-- change that to libHSpackagename-ghcjsversion_ghcversion
-- We need to do that in two places:
--    - Arguments for the system linker
--    - GHCi dynamic linker
ghcDynLibVersionTag :: String
ghcDynLibVersionTag   = "-ghc" ++ cProjectVersion

-- for now, Cabal installs libs with the GHC version
ghcjsDynLibVersionTag :: String
ghcjsDynLibVersionTag = "-ghcjs" ++ cProjectVersion

ghcjsPackageHsLibs :: DynFlags -> PackageConfig -> [String]
ghcjsPackageHsLibs dflags p = map fixLib (packageHsLibs dflags' p)
  where
    dflags' = dflags { ways = filter (not . isWayCustom) (ways dflags) }
    isWayCustom (WayCustom {}) = True
    isWayCustom _              = False
    fixLib lib | "HS" `isPrefixOf` lib &&
                 ghcDynLibVersionTag `isInfixOf` lib =
      replace ghcDynLibVersionTag ghcjsDynLibVersionTag lib
               | otherwise = lib

--------------------------------------------------------

replace :: Eq a => [a] -> [a] -> [a] -> [a]
replace from to xs = go xs
  where
    go [] = []
    go xxs@(x:xs)
      | from `isPrefixOf` xxs = to ++ go (drop (length from) xxs)
      | otherwise = x : go xs

