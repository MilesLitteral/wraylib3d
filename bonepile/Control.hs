{-# LANGUAGE AllowAmbiguousTypes #-}

module Manifest.Control where

import Manifest.Control.NetworkManager
import Manifest.Control.UDP
