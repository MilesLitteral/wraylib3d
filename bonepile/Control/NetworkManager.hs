{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE BangPatterns #-}
-- {-# LANGUAGE DatatypeContexts #-}

module Manifest.Control.NetworkManager where 

-- import Manipipe.Client.Http ()
import Data.Aeson 
import Data.Aeson.KeyMap 
import Data.Maybe
import Data.Word 
import qualified Data.UUID as UD
import Data.Map.Strict
import qualified Data.Text as T
import Control.Concurrent
import Control.Monad.IO.Class
import Control.Exception(catch, SomeException)

import Control.Lens hiding ((.=))
import Manipipe.Util.Network
import Manifest.Utils.Log
import Numeric (showHex)

import qualified Data.ByteString as B
import qualified Data.ByteString.Char8 as BC
import qualified Data.ByteString.Lazy as BL
import qualified Data.ByteString.Lazy.Char8 as LC
import qualified Data.ByteString.Char8 as B8
import qualified Data.Time.Clock as T
import qualified Data.Time.Clock.System as T

-- Make a Rule for checking if camera is available, apply to all endpoints
-- GET      /camera             -- List default configuration settings (shutterspeed, aperture, iso)
-- GET      /camera/{setting}   -- Returns configuration setting for setting
-- POST     /camera             -- Opens connection to camera
-- DELETE   /camera             -- Closes connection to camera
-- PUT      /camera             -- Sets all configurations listed in request body

-- POST     /preview            --Transfers image preview to memory
-- POST     /preview/all        --Transfers previews for all images on camera
-- GET      /preview/{name}     --Returns preview for image with given name

-- POST     /image              --Transfers all specified images to memory
-- GET      /image/list         --Lists out all images saved on camera memory card
-- GET      /image/{name}       --Returns the image with the given name
-- DELETE   /image              --Clears SD card on camera

-- POST     /trigger            --Presses down the cameras trigger button
-- DELETE   /trigger            --Releases cameras trigger button

-- POST     /capture            -- Captures an image on the camera
-- GET      /capture            -- Lists the number of captures on camera this session

-- GET      /ping               -- Sends OK as long as server is running
-- GET      /ping/camera        -- Sends OK if communication with camera still possible

-- data CommandType
--     = CommandTrigger
--     | CommandUnTrigger
--     | CommandInitialize
--     | CommandSelect
--     | CommandSelectBroadcast
--     | CommandUpload
--     | CommandDownload
--     | CommandDownloadBroadcast

-- class IsCommand command where
--     verify :: command -> CommandType

-- instance IsCommand String where
--     verify a = case a of
--         "HELLO" -> CommandTrigger

-- Attach a list of presets as the body of a Trigger event
-- The Pi will see the Preset and act accordingly,
-- Sending On/Off States to the lights on it's GPIO pins

data LightPreset
    = LSTATE0 -- 00 /0
    | LSTATE1 -- 01 /1
    | LSTATE2 -- 10 /4
    | LSTATE3 -- 11 /7
    deriving(Show, Eq)

data ServerResponse a =
    ValidServerResponse {srResponse     :: a}
  | ErrorServerResponse {srErrorMessage :: String}
  deriving (Eq, Show)

data HardwareCameraResponse = CameraResponse {
    camera ::  Int
    }  deriving (Eq, Show)

data HardwareConfigResponse 
    = ConfigResponse {
        aperture     :: String,
        iso          :: String,
        shutterspeed :: String
    }  deriving (Eq, Show)

data HardwareErrorResponse 
    = ErrorResponse {
        errorMessage :: String
    }  deriving (Eq, Show)

-- PTP(IEEE1588) CODE START
data EventPacketData =
    CamCaptureEventData {
      capTrigger  :: Bool,
      capUuid     :: UD.UUID
    }
  | CamFocusEventData
  | LightEventData {
      lightState  :: Int
    } deriving (Show)

data EventPacket = EventPacket{
    eTime  :: T.SystemTime,
    eData  :: EventPacketData
    } deriving (Show)

-- data EventPacket s = EventPacket{
--     eTime  :: T.SystemTime,
--     eState :: s,
--     eUUID  :: UD.UUID
--     } deriving (Show)

instance ToJSON EventPacket where
    toJSON (EventPacket st d) = do 
        object [ 
                ("seconds",      toJSON $ (showHex (T.systemSeconds    st) "")),
                ("microseconds", toJSON $ (showHex (systemMicroseconds st) "")),
                ("data",         toJSON d)
            ]

instance ToJSON EventPacketData where
    toJSON eventData = do 
        case eventData of 
            CamCaptureEventData s u-> object [ 
                                              ("type",  "capture"),
                                              ("state", toJSON s),
                                              ("uuid",  toJSON $ UD.toString u)
                                             ]
            CamFocusEventData -> object [("type", "focus")]
            LightEventData s -> object [ 
                                        ("type",  "light"),
                                        ("state", toJSON s)
                                       ]  
                                       
instance (FromJSON a) => FromJSON (ServerResponse a) where
    parseJSON (Object v) = do 
        result <- v .:? "response"
        case result of 
            Just res -> ValidServerResponse <$> parseJSON res
            Nothing  -> do  
                err <- v .:? "errorMessage"
                case err of
                    Just errMsg -> return $ ErrorServerResponse errMsg
                    Nothing -> return $ ErrorServerResponse "Response too malformed to get error message."
    parseJSON _ = return $ ErrorServerResponse "Cannot parse server response"
    
instance ToJSON HardwareCameraResponse where 
    toJSON s = object    [ "camera"     .= camera s ]

instance FromJSON HardwareCameraResponse where
    parseJSON (Object v) = CameraResponse <$> v .: "camera"

instance ToJSON   HardwareConfigResponse where 
    toJSON s = object    [ "aperture"          .= aperture s,
                            "iso"              .= Manifest.Control.NetworkManager.iso s, 
                            "shutterspeed"     .= shutterspeed s]

instance FromJSON HardwareConfigResponse where
    parseJSON (Object v) = do
        resp     <- v    .: "response" 
        aperture <- resp .: "aperture"
        iso      <- resp .: "iso"
        shutter  <- resp .: "shutterspeed"
        return $ ConfigResponse aperture iso shutter
     
instance ToJSON HardwareErrorResponse where 
    toJSON s = object    [ "errorMessage"     .= Manifest.Control.NetworkManager.errorMessage s ]
    
instance FromJSON HardwareErrorResponse where
    parseJSON (Object v) = ErrorResponse <$> v .: "errorMessage"

systemMicroseconds :: T.SystemTime -> Word32
systemMicroseconds (T.MkSystemTime _s ns) = floor $ fromIntegral ns / 1000.0

addMsToSystemTime :: T.SystemTime -> Word32 -> T.SystemTime
addMsToSystemTime (T.MkSystemTime s ns) ms = 
    let new_ns = ns + (ms * 1000000)
    in T.MkSystemTime s new_ns

sendSequence ::  [(String, [EventPacket])] -> IO [String] --[(Network.Stream.Result (Response String))]--String))]
sendSequence urls = mapM (\x -> do
    urlInfo <- urlInfoFromString ("http://" ++ (fst x) ++ "/events") 
    let events :: Data.Map.Strict.Map T.Text [EventPacket] = Data.Map.Strict.fromList [("events", (snd x))]
    --orderedMessageS MANI_LOG_BODY ("PTP Events to be sent(raw request): " ++ (show $ encode events)) 
    eResponse <- catch (Right <$> requestWithJson asIgnore POST urlInfo events) (\(_ :: SomeException)-> return $ Left "It broke")
    case eResponse of 
        Left msg -> return msg 
        Right () -> return $ (fst x) ++ "sequence sent"
    ) urls
-- PTP(IEEE1588) SPECIFIC CODE END

extractServerResponse :: forall a . (FromJSON a) => String -> ServerResponse a
extractServerResponse respStr =
    let response = decode (BL.fromStrict $ BC.pack respStr) :: Maybe (ServerResponse a) in
    case response of
        Just resp -> resp
        Nothing   -> ErrorServerResponse "Server response parsing error."

makeServerIgnoreRequest :: IO () -> IO (ServerResponse ())
makeServerIgnoreRequest request = do
    -- catches HTTP errors and inserts them into a ServerResponse
    -- In the future it will filter HTTP status code errors to make error messages more clear.
    response <- catch (Right <$> request) (\(e::SomeException)-> return $ Left $ show e)
    case response of
        Left msg -> return $ ErrorServerResponse ("Request failed: " ++ msg)
        Right _ -> return $ ValidServerResponse ()

makeServerByteStringRequest :: IO B.ByteString -> IO (ServerResponse B.ByteString)
makeServerByteStringRequest request = do
    -- catches HTTP errors and inserts them into a ServerResponse
    -- In the future it will filter HTTP status code errors to make error messages more clear.
    response <- catch (Right <$> request) (\(e::SomeException)-> return $ Left $ show e)
    case response of
        Left msg -> return $ ErrorServerResponse ("Request failed: " ++ msg)
        Right bytes -> return $ ValidServerResponse bytes

makeServerJSONRequest :: (FromJSON a) => IO (ServerResponse a) -> IO (ServerResponse a)
makeServerJSONRequest request = do
    -- catches HTTP errors and inserts them into a ServerResponse
    -- In the future it will filter HTTP status code errors to make error messages more clear.
    response <- catch (Right <$> request) (\(e::SomeException)-> return $ Left $ show e)
    case response of
        Left msg -> return $ ErrorServerResponse ("Request failed: " ++ msg)
        Right (serverResponse :: ServerResponse a) -> 
            case serverResponse of
                ErrorServerResponse msg -> return $ ErrorServerResponse ("Server returned error: " ++ msg)
                ValidServerResponse resp -> return $ ValidServerResponse resp
 
initializeCameraServers ::  [String] -> IO [String]
initializeCameraServers urls = mapM (\host -> do
    url <- urlInfoFromString ("http://" ++ host ++ "/camera")
    (response :: ServerResponse String) <- makeServerJSONRequest $ requestWithNone asJSON POST url
    case response of 
        ValidServerResponse responseStr -> return responseStr
        ErrorServerResponse msg -> do
            putStrLn msg
            return msg
    ) urls

pingCameraServers   :: [String] -> IO [Int]
pingCameraServers urls  = mapM (\host -> do
    url <- urlInfoFromString ("http://" ++ host ++ "/ping")
    (eResponse :: ServerResponse (Data.Map.Strict.Map T.Text Int)) <- makeServerJSONRequest $ requestWithNone asJSON GET url
    case eResponse of 
        ValidServerResponse responseJson -> return $ responseJson ! "camera"
        ErrorServerResponse msg -> do
            putStrLn msg
            return (-1)     
    ) urls


pingCameraDirect   :: [String] -> IO [Int]
pingCameraDirect urls  = mapM (\host -> do
    url <- urlInfoFromString ("http://" ++ host ++ "/ping/camera")
    (response :: ServerResponse (Data.Map.Strict.Map T.Text Int)) <- makeServerJSONRequest $ requestWithNone asJSON GET url
    case response of 
        ValidServerResponse responseJson -> return $ responseJson ! "camera"
        ErrorServerResponse msg -> do
            putStrLn msg
            return (-1)
    ) urls

getCameraServersConfig   :: String -> IO [(String, [String])]
getCameraServersConfig camHost  = do
    let defaultConfigs = ["shutterspeed", "aperture", "iso", "autopoweroff"]
    let unavailable    = "Unavailable"
    print camHost 
    url <- urlInfoFromString ("http://" ++ camHost ++ "/camera")
    (response :: ServerResponse (Data.Map.Strict.Map T.Text String)) <- makeServerJSONRequest $ requestWithNone asJSON GET url
    activeCfgs <- case response of
        ValidServerResponse responseJson -> return responseJson
        ErrorServerResponse msg -> do
            putStrLn msg
            return Data.Map.Strict.empty
    url <- urlInfoFromString ("http://" ++ camHost ++ "/camera?config=yes")
    (response :: ServerResponse (Data.Map.Strict.Map T.Text [String])) <- makeServerJSONRequest $ requestWithNone asJSON GET url
    cfgOptions <- case response of 
        ValidServerResponse responseJson -> return responseJson
        ErrorServerResponse msg -> do
            putStrLn msg
            return Data.Map.Strict.empty
    return $ Prelude.map (\settingName -> 
        let active  = findWithDefault unavailable settingName activeCfgs
            options = findWithDefault [] settingName cfgOptions
        in (active, options)) defaultConfigs

setCameraServersConfigSetting   :: [String] -> HardwareConfigResponse -> IO ()
setCameraServersConfigSetting  urls newConfig = mapM_ (\host -> do
    url <- urlInfoFromString ("http://" ++ host ++ "/camera")
    response <- makeServerIgnoreRequest $ requestWithJson asIgnore PUT url newConfig
    case response of 
        ErrorServerResponse msg -> do 
            putStrLn msg
        ValidServerResponse _ -> return ()
    ) urls

deleteCameraServers :: [String] -> IO [Int]--[(Network.Stream.Result    (Response String))]
deleteCameraServers urls = mapM (\host -> do
    url <- urlInfoFromString ("http://" ++ host ++ "/camera")
    (response :: ServerResponse (Data.Map.Strict.Map T.Text Int)) <- makeServerJSONRequest $ requestWithNone asJSON DELETE url
    case response of 
        ValidServerResponse responseJson -> return $ responseJson ! "camera"
        ErrorServerResponse msg -> do
            putStrLn msg
            return (-1)
    ) urls

triggerCameraServers :: [String] -> IO [Int] --[(Network.Stream.Result (Response String))]
triggerCameraServers urls = mapM (\host -> do
    url <- urlInfoFromString ("http://" ++ host ++ "/trigger")
    (response :: ServerResponse (Data.Map.Strict.Map T.Text Int)) <- makeServerJSONRequest $ requestWithNone asJSON POST url
    case response of 
        ValidServerResponse responseJson -> return $ responseJson ! "camera"
        ErrorServerResponse msg -> do
            putStrLn msg
            return (-1)
    ) urls

-- triggerCameraServersWithLightStates :: [String] -> [LightPreset] -> IO [Int] --[(Network.Stream.Result (Response String))]
-- triggerCameraServersWithLightStates urls lights = mapM (\x -> do
--     urlInfo <- urlInfoFromString ("http://" ++ x ++ "/trigger")  
--     eResponse <- catch (Right <$> requestWithNone asJSON POST <$> urlInfo lights) (\(_ ::SomeException)-> return $ Left "It broke")
--     case eResponse of 
--         Left _ -> return (-1)
--         Right (response :: IO (ServerResponse (Data.Map.Strict.Map T.Text Int))) -> do
--             response2 <- response
--             case response2 of 
--                 ValidServerResponse responseJson -> return $ responseJson ! "camera"
--                 ErrorServerResponse msg -> do
--                     putStrLn $ "Server error: " ++ msg
--                     return (-1)
--     ) urls
    
untriggerCameraServers :: [String] -> IO [Int] --[(Network.Stream.Result (Response String))]
untriggerCameraServers urls = mapM (\host -> do
    url <- urlInfoFromString ("http://" ++ host ++ "/trigger")
    (response :: ServerResponse (Data.Map.Strict.Map T.Text Int)) <- makeServerJSONRequest $ requestWithNone asJSON DELETE url
    case response of 
        ValidServerResponse responseJson -> return $ responseJson ! "camera"
        ErrorServerResponse msg -> do
            putStrLn msg
            return (-1)
    ) urls

getAllPreviews :: String -> IO [(String, B.ByteString)]
getAllPreviews host = do
    -- post preview all
    url <- urlInfoFromString ("http://" ++ host ++ "/preview/all")
    (response :: ServerResponse (Data.Map.Strict.Map T.Text [T.Text])) <- makeServerJSONRequest $ requestWithNone asJSON POST url
    imageList <- case response of 
        ValidServerResponse responseJson -> return $ responseJson ! "images"
        ErrorServerResponse msg -> do
            putStrLn msg
            return []
    mapM (\imageName -> do
        url <- urlInfoFromString ("http://" ++ host ++ "/image/" ++ T.unpack imageName)
        response <- makeServerByteStringRequest $ requestWithNone asByteString GET url
        case response of 
            ErrorServerResponse msg   -> error msg
            ValidServerResponse bytes -> return (T.unpack imageName, bytes) 
        ) imageList

getAllCaptures :: [(String, [UD.UUID])] -> Bool -> IO (Data.Map.Strict.Map UD.UUID B.ByteString)
getAllCaptures desiredCaptures preview = do
    let endpoint = if preview then "/preview" else "/image"
    transferReqResults <- mapM (\(host, desiredUuids)-> do
        url <- urlInfoFromString ("http://" ++ host ++ endpoint)
        let payload = Data.Map.Strict.fromList [("captures"::T.Text, Prelude.map UD.toString desiredUuids)]
        response <- makeServerIgnoreRequest $ requestWithJson asIgnore POST url payload
        case response of
            ErrorServerResponse msg -> do
                putStrLn msg
                return False
            ValidServerResponse _ -> return True
        ) desiredCaptures
    downloadedImages <- mapM (\(host, desiredUuids)-> do
        Data.Map.Strict.fromList <$> mapM (\uuid -> do
                url <- urlInfoFromString ("http://" ++ host ++ "/image/" ++ UD.toString uuid)
                response <- makeServerByteStringRequest $ requestWithNone asByteString GET url
                case response of
                    ErrorServerResponse msg -> error msg
                    ValidServerResponse img -> return (uuid, img)
            ) desiredUuids
        ) desiredCaptures
    return $ Data.Map.Strict.unions downloadedImages

downloadCameraServersPreviewContents :: [String] -> IO [Int] --[(Network.Stream.Result (Response String))]
downloadCameraServersPreviewContents urls = mapM (\host -> do
    url <- urlInfoFromString ("http://" ++ host ++ "/trigger")
    (response:: ServerResponse (Data.Map.Strict.Map T.Text Int)) <- makeServerJSONRequest $ requestWithNone asJSON POST url
    case response of 
        ValidServerResponse responseJson -> return $ responseJson ! "camera"
        ErrorServerResponse msg -> do
            putStrLn msg
            return (-1)
    ) urls
    
downloadCameraSelectServersContents :: [String] -> IO [Int] --[(Network.Stream.Result (Response String))]
downloadCameraSelectServersContents urls = mapM (\host -> do
    url <- urlInfoFromString ("http://" ++ host)
    (response:: ServerResponse (Data.Map.Strict.Map T.Text Int)) <- makeServerJSONRequest $ requestWithNone asJSON GET url
    case response of 
        ValidServerResponse responseJson -> return $ responseJson ! "camera"
        ErrorServerResponse msg -> do
            putStrLn msg
            return (-1)
    ) urls

downloadCameraServersContents :: String -> IO [(String, B.ByteString)] --[(Network.Stream.Result (Response String))]
downloadCameraServersContents url = do
    -- post preview all
    eResponse <- catch (Right <$> requestWithNone asJSON GET <$> urlInfoFromString ("http://" ++ url ++ "/image/list")) (\(e::SomeException)-> return $ Left $ "Preview Error: " ++ show e)
    imageList <- case eResponse of 
        Left msg -> error msg
        Right (response :: IO (ServerResponse (Data.Map.Strict.Map T.Text [T.Text]))) -> do
            response2 <- response
            case response2 of
                ValidServerResponse responseJson -> do
                    --print  $ responseJson
                    return $ responseJson ! "names"
                ErrorServerResponse msg -> do
                    putStrLn $ "Server error: " ++ msg
                    return []
    -- -- post imageList to /image/
    postUrl <- urlInfoFromString (url ++ "/image")
    let postJSON = Data.Map.Strict.fromList [("images", imageList)] :: Data.Map.Strict.Map T.Text [T.Text]
    -- -- end createJSON Body
    !ePostResponse <- catch (Right <$> requestWithJson asIgnore POST postUrl postJSON) (\(e::SomeException)-> return $ Left $ "Preview Error: " ++ show e)
    case ePostResponse of 
        Left msg -> error msg
        Right _ -> do
            print $ "Request sent"
            threadDelay 100000
            return ()
    mapM (\imageName -> do
        eImageResponse <- catch (Right <$> requestWithNone asByteString GET <$> urlInfoFromString ("http://" ++ url ++ "/image/" ++ (T.unpack imageName))) (\(e::SomeException)-> return $ Left $ show e)
        print $ "Response Received"
        case eImageResponse of 
            Left msg -> error "Bad Images Received" --return ("", B.empty)
            Right bytes -> do 
                print $ "Serializing Response"
                b <- (bytes :: IO B.ByteString)
                --B.writeFile ("./assets/rawImages/test.jpg") b
                print "Return Result"
                return (T.unpack imageName, b) 
        ) imageList

formatCameraServersContents   :: [String] -> IO [Int] --[(Network.Stream.Result (Response String))]
formatCameraServersContents   urls = mapM (\host -> do
    url <- urlInfoFromString ("http://" ++ host ++ "/image")
    (response :: ServerResponse (Data.Map.Strict.Map T.Text Int)) <- makeServerJSONRequest $ requestWithNone asJSON DELETE url
    case response of 
        ValidServerResponse responseJson -> return $ responseJson ! "camera"
        ErrorServerResponse msg -> do
            putStrLn msg
            return (-1)
    ) urls

closeCameraServers :: [String] -> IO [Int] --[(Network.Stream.Result (Response String))]
closeCameraServers urls = mapM (\host -> do
    url <- urlInfoFromString ("http://" ++ host ++ "/camera")
    (response :: ServerResponse (Data.Map.Strict.Map T.Text Int)) <- makeServerJSONRequest $ requestWithNone asJSON DELETE url
    case response of 
        ValidServerResponse responseJson -> return $ responseJson ! "camera"
        ErrorServerResponse msg -> do
            putStrLn msg
            return (-1)
    ) urls

    