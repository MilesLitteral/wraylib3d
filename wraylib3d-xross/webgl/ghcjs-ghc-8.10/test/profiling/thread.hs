import Control.Concurrent
import Control.Concurrent.MVar
import Control.Monad
import GHC.Stack

f1 :: IO [String]
f1 = f2 >>= return

f2 :: IO [String]
f2 = currentCallStack

main1 v ret = do
    putMVar ret . show =<< f1
    putMVar v ()

someF :: Int -> IO String
someF n = replicateM n (return '.')

main2 v ret = do
    putMVar ret . show =<< whoCreated =<< someF 10
    putMVar v ()

wait [] = return ()
wait vs@(v : r) = do
    readMVar v
    wait r

main = do
    var1 <- newEmptyMVar
    var2 <- newEmptyMVar

    ret1 <- newEmptyMVar
    ret2 <- newEmptyMVar

    i1 <- forkIO (main1 var1 ret1)
    i2 <- forkIO (main2 var2 ret2)

    wait [var1, var2]
    putStrLn =<< readMVar ret1
    putStrLn =<< readMVar ret2
