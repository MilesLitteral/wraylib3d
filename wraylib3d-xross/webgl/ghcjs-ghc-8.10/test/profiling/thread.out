["Main.main (thread.hs:(28,1)-(40,30))","Main.main1 (thread.hs:(12,1)-(14,16))","Main.f1 (thread.hs:7:1-18)"]
["Main.main (thread.hs:(28,1)-(40,30))","Main.main2 (thread.hs:(19,1)-(21,16))","Main.someF (thread.hs:17:1-35)"]
